includes/class-sensei-quiz.php

```diff
@@ -672,61 +672,71 @@ private static function is_uploaded_file_valid( $file_path, $file_name, $questio
 		if ( empty( $lesson_id ) || empty( $user_id )
-			|| 'lesson' != get_post_type( $lesson_id )
+			|| 'lesson' !== get_post_type( $lesson_id )
 			|| ! get_userdata( $user_id ) ) {
 			return false;
 		}
 
-		// get the users lesson status to make
+		// Get the users lesson status to make.
 		$user_lesson_status = Sensei_Utils::user_lesson_status( $lesson_id, $user_id );
 		if ( ! isset( $user_lesson_status->comment_ID ) ) {
-			// this user is not taking this lesson so this process is not needed
+			// This user is not taking this lesson so this process is not needed.
 			return false;
 		}
 
-		// get the lesson quiz and course
+		// Get the lesson quiz and course.
 		$quiz_id   = Sensei()->lesson->lesson_quizzes( $lesson_id );
 		$course_id = Sensei()->lesson->get_course_id( $lesson_id );
 
-		// reset the transients
+		// Reset the transients.
 		$answers_transient_key          = 'sensei_answers_' . $user_id . '_' . $lesson_id;
 		$grades_transient_key           = 'quiz_grades_' . $user_id . '_' . $lesson_id;
 		$answers_feedback_transient_key = 'sensei_answers_feedback_' . $user_id . '_' . $lesson_id;
 		delete_transient( $answers_transient_key );
 		delete_transient( $grades_transient_key );
 		delete_transient( $answers_feedback_transient_key );
 
-		// reset the quiz answers and feedback notes
+		// Reset the quiz answers and feedback notes.
 		Sensei_Utils::delete_user_data( 'quiz_answers', $lesson_id, $user_id );
 		Sensei_Utils::delete_user_data( 'quiz_grades', $lesson_id, $user_id );
 		Sensei_Utils::delete_user_data( 'quiz_answers_feedback', $lesson_id, $user_id );
 
-		// Delete quiz answers, this auto deletes the corresponding meta data, such as the question/answer grade
+		// Delete quiz answers, this auto deletes the corresponding meta data, such as the question/answer grade.
 		Sensei_Utils::sensei_delete_quiz_answers( $quiz_id, $user_id );
 
-		Sensei_Utils::update_lesson_status(
-			$user_id,
-			$lesson_id,
-			'in-progress',
-			array(
+		$lesson_progress_repository = Sensei()->lesson_progress_repository_factory->create();
+		$lesson_progress            = $lesson_progress_repository->get( $lesson_id, $user_id );
+		if ( $lesson_progress ) {
+			$lesson_progress->start();
+			$lesson_progress_repository->save( $lesson_progress );
+
+			// Save updated metadata.
+			$metadata = [
 				'questions_asked' => '',
 				'grade'           => '',
-			)
-		);
+			];
+			foreach ( $metadata as $key => $value ) {
+				update_comment_meta( $lesson_progress->get_id(), $key, $value );
+			}
+		}
 
-		// Update course completion
-		Sensei_Utils::update_course_status( $user_id, $course_id );
+		// Update course completion.
+		$course_progress_repository = Sensei()->course_progress_repository_factory->create();
+		$course_progress            = $course_progress_repository->get( $course_id, $user_id );
+		if ( $course_progress ) {
+			$course_progress->start();
+			$course_progress_repository->save( $course_progress );
```

Comment:
Would it make sense for saving to be implicit instead, when calling `$course_progress->start();` and similar methods? Or is there a benefit of working with a state only in memory before persisting it? Are there many cases where we do multiple changes on a progress object?

Reply:
_Dmitry did most of the architecture but he is AFK right now so I will try to paraphrase his point of view._ 🙂 

I think the idea here is to have a separation of concerns. The models are only responsible for storing the state and the repositories for interacting with the database. The advantages of this approach are:

* This keeps things very simple, straightforward, and easy to reason about.
* There is no interdependency between models and repositories.
* It's easier to test.
* It's consistent with what we did in sensei-pro for the [student groups feature](https://github.com/Automattic/sensei-pro/tree/trunk/modules/student-groups/includes).

It's funny that you mention this because I also like fluent interfaces (Laravel fanboy here) and I've asked Dmitry the same question. 😄 But now I can see why he did it this way.

Please let me know if you have strong opinion about this and we could discuss it further.

Reply:
It's great to separate the model and repositories however this makes the API a bit hard to use. Maybe we should introduce a facade that would hide the implementation details and make each operation a single call?

Reply:
Yes, a few higher level helpers would be a great place for a friendlier interface. For example, methods like
`Sensei_Progress::start_course( $course_id, $user_id = get_current_user_id() )`
`Sensei_Progress::complete_lesson( $lesson_id )`
or 
`Sensei\Lesson_Progress::complete( $lesson_id )`

Reply:
This is a great suggestion, but I think it's out of scope for this PR. As mentioned in [this comment](https://github.com/Automattic/sensei/pull/5428#issuecomment-1234929717), the goal of this PR is to create an API for the _data_ layer (lower level) and not to refactor the _logic_ layer (higher level). So I think it's ok to keep using `Sensei_Utils::user_start_course` for now and refactor that in a future PR.